### PR TITLE
Create scaffold for Angular 6 client (WIP)

### PR DESCRIPTION
**Pull request by tsiegleauq**
My branch is based on Master, this one seems older. Please update OpenSlides/OpenSlides-3 first

- basic client structure
- authentication towards OS-server
- access REST-Api
- webpack proxy for convenient usage
- fontawesome
- twitter bootstrap 3.3.7
- login example (recreate login page)
- routing
- template for agenda and motions
- dynamic content loading
- custom alert component

### CODE REVIEW

**Comment by FinnStutzenstein** [comment]
@tsiegleauq I updated the `OpenSlides/OpenSlides-3` branch to the current master. There was one big feature added. This should be in OS3. Please rebase your branch on the `OpenSlides/OpenSlides3` branch.

**Comment by FinnStutzenstein** [comment]
Yes, these components should go into an own folder. I was thinking about not to put everything into the app/ folder.
Whats about creating an site/alert/ folder for alerts, /site/start/, ..., and also move the motions/, agenda/, ... into the site/ folder. So the hirachial structure is kept along with the routing?

**Comment by tsiegleauq** [comment]
> Whats about creating an site/alert/ folder for alerts, /site/start/, ..., and also move the motions/, agenda/, ... into the site/ folder. So the hirachial structure is kept along with the routing?

Would be fine. I was discussing this with @emanuelschuetze on friday as well, at some point we need a structure like 

`/site/motions/motion-categories/motion-category.ts/` (would be)
would be this list: https://demo.openslides.org/motions/category
`/site/motions/motion-categories/motion-category-details/motion-category-detail.ts`
for the details.

(pluralisation is another topic we might discuss. eg. _motion/motions_ _category/categories_, )

**Comment by FinnStutzenstein** [comment]
Login will be moved into a separate folder
4 spaces is correct,  has to be changed in the config ;)
The OpenSlides Service will handle the startup and put all bootstrap things together. The name is ok, because this will have the "heart" of the code later on.
The title service will be expanded later.
The "core" folder is the "shared" one from the doku. You can rename it if you want. Then we should create subfolders for each service, component, .. but do not classify by type
Yes, I did forget the subscribe...

**Comment by tsiegleauq** [comment]
`4 spaces is correct, has to be changed in the config ;)`
2 spaces are default in angular 6.
please adjust to the coding style you are committing to.

https://github.com/angular/angular-cli/issues/3219

> The OpenSlides Service will handle the startup and put all bootstrap things together. The name is ok, because this will have the "heart" of the code later on.

I cannot really image what you plan to achieve with that. Keep in mind that services are supposed to provide functionality to components and are inherently singletons. 

`Then we should create subfolders for each service, component, .. but do not classify by type`

**Comment by FinnStutzenstein** [comment]
Hey, there are currently two things wip:
- the auth system (as you see because the not working logout). So I didnt test anything yet.
- getting permissions from the server. This should be the next step, because permissions are everywhere heavily used, e.g. for the auth guard.

So the codebase isn't perfect, but wasn't before, too. And the plan for the OpenSlides Service: see the start.js in the current codebase. This basic startup logic needs to be placed somewhere. It will initalize the authentication, connect the Websocket and will handle the first update of the websocket cache (when it will be implemented).

**Comment by tsiegleauq** [comment]
> This should be the next step, because permissions are everywhere heavily used, e.g. for the auth guard.

I disagree. 
Implementing permissions now would mean to make the third step before the second.
At the current point of the development, permissions can not be implemented "the right way".
Software is supposed to grow organically, feature by feature.
Implement permission makes sense when the software is mature enough to actually behave differently with different permissions (again, I do not want to port all the issues of the current version of the new client)

> So the codebase isn't perfect, but wasn't before, too.

"perfection" is subjective and surely discussable. In my opinion, the code was cleaner before. What is not discussable is distinct coding style ;-)

**Comment by tsiegleauq** [comment]
I tried out you example and, interestingly enough, routing is not working as expected anymore. `/login` now is nested into site-component. It was not, and never, defined that way.

I know you want the new client to behave like the old one, which is fine, but not for now. Also, with that practice, you will likely (or definitely) import all the anti-pattern of the old software into the new one.
Do not take too many steps at once, keep the components distinct as long as possible, keep features simple (always), let software grow and, most importantly, do not force functionality into a software before the functionality is necessary.

**Comment by tsiegleauq** [comment]
That being said, I will close this pull request and abandon the changes.
Changes made here will, in part, find their way into a new independent Repository dedicated to the OpenSlides3 Angular 6 client.
The new client will probably find it's way into OpenSlides at some point.

There are now serveral reasons for that:
* keeping a branch up to date is an unnecessary overhead
* I need to be able to review commits to the Angular6 Client to make sure the Guidlines are kept
* developing Client and Server independently is no problem at all, there is really no reason to keep the whole OpenSlides stack, since the client is only dependent on the REST-API calls.
* the code quality suffered a lot after the last commit and the regressions are too severe

Thank you for future participation in [OpenSlides3-Client](https://github.com/tsiegleauq/OpenSlides3-Client)  @FinnStutzenstein.

**Comment by ThomasJunk** [comment]
I think this is a healthy decision. 

If you have the chance to have something like a restart, it is nice to define common coding guidelines. 

I welcome the decision to split the redesign off the official repo. 
 
:+1:

**Comment by tsiegleauq** [comment]
Changes mention above were made cleaned up, refactored and merged together.
For the time being it was decided to continue OpenSlides-3 here.

@FinnStutzenstein @emanuelschuetze please review changes.

File: client/src/app/core/services/websocket.service.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        if (!this.subject) {
+            this.subject = webSocket(socketProtocol + socketServer + socketPath);
+        }
+        return this.subject;
+    }
+
+    // delegates to websockets for either the side or projector websocket
+    private getWebSocketPath(): string {
+        const currentRoute = this.router.url;
+        if (currentRoute.includes('projector')) {
```
Comment (FinnStutzenstein) [inline client/src/app/core/services/websocket.service.ts]: A little note: Can you make this more specific. We later may have routes like `/manage-projectors/`, where this check should be false. E.g. explicit a startsWith with `/projector/` and `/real-projector/`.

**Comment by tsiegleauq** [comment]
@FinnStutzenstein 
I made some slight adjustments. Feel free to continue.

File: client/src/app/site/site.component.ts [hunk 2]
```diff
@@ -45,6 +48,20 @@ export class SiteComponent implements OnInit {
         socket.next(val => {
             console.log('socket.next: ', val);
         });
+
+        //get a translation via code: use the translation service
+        this.translate.get('Motions').subscribe((res: string) => {
```
Comment (FinnStutzenstein) [inline client/src/app/site/site.component.ts]: @tsiegleauq Is there a possibility to get the result synchronously? We need the translations very often in the code, and doing everythin translationrelated asynchronly will result in much more complex code.

**Pull request merged**